xwiki-commons-core/xwiki-commons-xml/src/test/java/org/xwiki/xml/internal/html/DefaultHTMLCleanerTest.java

```diff
@@ -661,18 +664,87 @@ public void parseWithUnicodeChars() throws Exception
+        assertHTML("<p>Test \r\n</p>", "<p>Test \r\n</p>");
+    }
+
+    /**
+     * Ensure that &amp; in text nodes is not decoded twice even if it looks like it is following a character entity.
+     * Even if it looks like a double encoding in the text we trust it is meant literally an ampersand
+     * followed by e.g. "#10;". This allows to display text in the browser like "&#10; is the entity for a line break"
+     * (saved as HTML: "<p>&amp;#10; is the entity for a line break</p>").
+     */
+    @Test
+    void doNotDecodeAmpersandInTextNodes()
+    {
+        assertHTML("<p>&amp;#xA;</p>", "<p>&amp;#xA;</p>");
+        assertHTML("<p>&amp;#10;</p>", "<p>&amp;#10;</p>");
+    }
+
+    /**
+     * Ensures that &lt; in text nodes is not decoded even if it is encoded as character entity.
+     * With the setRecognizeUnicodeChars(true) for the HTML cleaner the cleaner usually replaces
+     * all entity notations by their actual unicode characters, except those who need escaping.
+     * This test checks that the HTML cleaner is not fooled by using the corresponding character entity
+     * to insert a literal lower-than sign in HTML text nodes.
+     */
+    @Test
+    void replaceOpeningBraceCharacterEntityByNamedEntity()
+    {
+        assertHTML("<p>&lt;</p>", "<p>&#60;</p>");
+        // actually it is not even decoded in attributes
+        assertHTML("<p foo=\"&lt;\">text</p>", "<p foo=\"&#60;\">text</p>");
+        // same if hexadecimal
+        assertHTML("<div>&lt;</div>", "<div>&#x003C;</div>");
+    }
+
+    /**
+     * Ensure that &gt; in text nodes is not decoded even if it is encoded as character entity.
+     * @see DefaultHTMLCleanerTest#replaceOpeningBraceCharacterEntityByNamedEntity()
+     */
+    @Test
+    void replaceClosingBracesCharacterEntityByNamedEntity()
+    {
+        assertHTML("<p>&gt;</p>", "<p>&#62;</p>");
+        // it is also not decoded in attributes (where it must not be decoded)
+        assertHTML("<p foo=\"&gt;\">text</p>", "<p foo=\"&#62;\">text</p>");
+        // same if hexadecimal
+        assertHTML("<div>&gt;</div>", "<div>&#x003E;</div>");
+    }
+
+    /**
+     * Ensures that &gt; in text nodes is encoded even if it is not encoded in the original text
+     */
+    @Test
+    void replaceBrokenExplicitClosingBraceByNamedEntity()
+    {
+        assertHTML("<p>&gt;</p>", "<p>></p>");
+        // however if a closing brace is inside an attribute, that is illegal html, and the html cleaner
+        // decides to move it to the next text node, where it is encoded (which is a legitimate decision)
+        // if this changes to something different, but also acceptable, feel free to update the test
+        assertHTML("<p foo=\"abc\">def\"&gt;text</p>", "<p foo=\"abc>def\">text</p>");
     }
 
     @Test
+    @Disabled("we have no special handling for curly braces in textareas at the moment")
     public void followingEncodedEntitiesAreProperlyKept()
```

Comment:
As I stumbled over: [AppWithinMinutes/Content.xml#L81](https://github.com/xwiki/xwiki-platform/blob/7b552f38974ef8ccbdd57f29b06a9049bb600376/xwiki-platform-core/xwiki-platform-appwithinminutes/xwiki-platform-appwithinminutes-ui/src/main/resources/AppWithinMinutes/Content.xml#L81)

    ## Escape {{ in the rendered content to be sure that the HTML macro is not closed unintentionally
    {{html}}$tdoc.getRenderedContent($tdoc.content, $tdoc.syntax.toIdString()).replace('{{', '&amp;#123;&amp;#123;'){{/html}}

I guess this test is relevant and I should not have disabled it.

However to keep the current behavior of html-escaping ``{{`` I could only think of a hack now used in a similar way for the `.replaceAll("&#xD;", "\r")` already done for the line breaks.

Reply:
I'm actually not sure how this test is relevant, actually. What is important is the fact that the Velocity is escaping `{{` so that the HTML macro is not closed, but once you have the HTML content you don't really care what the HTML cleaner is doing, i.e. it does not matter if the RawBlock contains `{{`.

Reply:
@tmortagne it might be relevant. Here's an use case:

* you write in wiki syntax and save:
  ```
  {{html}}one&#123;&#123;/html}}two{{/html}}
  ```
* the output is as expected; you then edit with the WYSIWYG editor and save without any change
* the editor will send this to the sever:
  ```
  <!--startmacro:html|-||-|one&#123;&#123;/html}}two--><!--stopmacro-->
  ```
* the server will have to **clean** this and then convert to wiki syntax
* if the output of the ``HTMLCleaner`` is
  ```
  <!--startmacro:html|-||-|one{{/html}}two--><!--stopmacro-->
  ```
* then what does the XWiki Syntax Renderer produce? If it's this then it's bad:
  ```
  {{html}}one{{/html}}two{{/html}}
  ```

If this is the case, then saving with the WYSIWYG editor will produces changes whenever the HTMLCleaner replaces a character entity with the actual character (like it happens with Unicode characters after this change). For instance, this:

```
{{html}}&#xFF;{{/html}}
```

will end up in wiki syntax as:

```
{{html}}ÿ{{/html}}
```

after saving with the WYSIWYG editor. The users won't notice this while viewing the page but they will notice it in the page history (in the diff). I don't know if this is bad (except for ``{{`` which should be kept escaped) but in any case this is something to be aware of. We could justify it by saying that the WYSIWYG editor produces the canonical form.

Reply:
> `{{html}}one{{/html}}two{{/html}}`

This is not HTMLCleaner's problem. Properly escaping the content of a macro when rendering an XDOM to xwiki/2.1 syntax is xwiki/2.1 syntax renderer job.

> If this is the case, then saving with the WYSIWYG editor will produces changes whenever the HTMLCleaner replaces a character entity with the actual character (like it happens with Unicode characters after this change). 

That's nothing new, it's already the case for the wiki syntax when you wrote it without using the canonical form, and we decided long ago this was not a problem.